### PR TITLE
App Service ACR Integration

### PR DESCRIPTION
**Pull request by erikschlegel**
## All Submissions:
-------------------------------------
* [YES] Have you followed the guidelines in our Contributing [document](./CONTRIBUTING.md)?
* [YES] Have you added an explanation of what your changes do and why you'd like us to include them?
* [NO] I have updated the documentation accordingly.
* [NA] I have added tests to cover my changes.
* [YES] All new and existing tests passed.
* [YES] My code follows the code style of this project.
* [YES] I ran lint checks locally prior to submission.
* [YES] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

## What is the current behavior?
-------------------------------------
Currently, our app service modules only support docker images that are publicly hosted in docker hub which will not work for a majority of our enterprise customers. 

Issue Number: #151 

## What is the new behavior?
-------------------------------------
This change proposal allows an app developer to integrate images hosted on a provisioned private container registry with a app service mesh. 

Here's some additional features included in this PR.

- Supports enabling continuous deployment for an app service.  
- Enhancement to azure simple to pull the reference image hosted on docker hub and deploy to ACR prior to app service provisioning.
- Integrates ACR build tasks to re-build the reference app image on each deployment. 
- Creates an ACR webhook to enable app service CD.
- Renamed the `azure-simple` template to `azure-paas-single-region`
- Renamed the `azure-simple-hw` template to `azure-hello-world`
- Upgraded to the latest version of terraform(0.12.1->0.12.2)
- Pushed the latest base image containing version 0.12.2 to the ACR sourced from our CI pipeline.

## Does this introduce a breaking change?
-------------------------------------
- [YES]

## Any relevant logs, error output, etc?
-------------------------------------
deployed docker container with CD enabled. 
image

## Other information
-------------------------------------

### CODE REVIEW

File: infra/templates/azure-paas-single-region/terraform.tfvars [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# GENERAL
+resource_group_location = "eastus"
+name = "cobalt-az-simple"
+# SERVICE PLAN
+app_service_name = {
+    cobalt-backend-api = "msftcse/cobalt-paas-single-region:release" 
+}
+# ACR image deploy source
+acr_build_git_source_url = "https://github.com/erikschlegel/echo-server.git"
```
Comment (nmiodice) [inline infra/templates/azure-paas-single-region/terraform.tfvars]: Any way to pull this into this repository? Would hate for this project to stop working if you delete this personal repo

File: infra/templates/azure-paas-single-region/tests/unit/template_unit_test.go [hunk 2]
```diff
@@ -35,8 +35,7 @@ func asMap(t *testing.T, jsonString string) map[string]interface{} {
 
 func TestTemplate(t *testing.T) {
 	expectedAppServicePlan := asMap(t, `{
-		"name":        "cobalt-backend-api-`+workspace+`",
-		"site_config": [{"linux_fx_version": "DOCKER|msftcse/cobalt-azure-simple:0.1"}]
```
Comment (nmiodice) [inline infra/templates/azure-paas-single-region/tests/unit/template_unit_test.go]: any reason to remove this?

File: README.md [hunk 3]
```diff
@@ -67,7 +67,7 @@ This pipeline is configured to trigger new builds for each new PR.
 
 ### Getting Started
 
-The easiest way to try Cobalt is to start with our [azure-simple](https://github.com/Microsoft/cobalt/tree/master/infra/templates/azure-simple-hw) template.
+The easiest way to try Cobalt is to start with our [azure-paas-single-region](https://github.com/Microsoft/cobalt/tree/master/infra/templates/azure-hellow-world) template.
```
[the remark this answers is not in the document]
Comment (nmiodice) [reply README.md]: discussed offline: `az-service-single-region`

File: infra/templates/azure-paas-single-region/appdev.tf [hunk 4]
```diff
@@ -0,0 +1,100 @@
+locals {
+  service_plan_name = "${local.prefix}-sp"
+  app_insights_name = "${local.prefix}-ai"
+  acr_webhook_name  = "cdwebhook"
```
[the remark this answers is not in the document]
Comment (erikschlegel) [reply infra/templates/azure-paas-single-region/appdev.tf]: Agree with adding the `prefix`. JFYI, webhooks prevents you from including special characters like `-` in the name.

File: infra/modules/providers/azure/app-service/main.tf [hunk 5]
```diff
@@ -1,6 +1,7 @@
 locals {
   access_restriction_description = "blocking public traffic to app service"
   access_restriction_name        = "vnet_restriction"
+  acr_webhook_name               = "cdwebhook"
```
Comment (nmiodice) [inline infra/modules/providers/azure/app-service/main.tf]: Should be unique per image as discussed in person

**Comment by nmiodice** [comment]
Ship it!

```
              |    |    |
             )_)  )_)  )_)
            )___))___))___)\
           )____)____)_____)\\
         _____|____|____|____\\\__
---------\                   /---------
  ^^^^^ ^^^^^^^^^^^^^^^^^^^^^
    ^^^^      ^^^^     ^^^    ^^
         ^^^^      ^^^

------------------------------------------------
```

**Pull request merged**